### PR TITLE
Add custom rendering example

### PR DESCRIPTION
**Pull request by rasmusgo**
This is a rebased and cleaned up version of #389.

### CODE REVIEW

**Review by kanerogers — CHANGES_REQUESTED** [review]
Overall, this looks really good! As an aside, you could experiment with hardware ray tracing here if your target is PCVR. You could definitely have a hardware optimised fast path for devices that support it, then on mobile do everything in software - you might be interested in some of arm's [ray tracing work](https://www.youtube.com/watch?v=47HxKPPHyeU).

My only concern here is the refactoring of the fragment shaders; getting some of these functions to take parameters instead of just reading from registers makes sense from a code re-use point of view, but if you're making changes to the fragment shader just to support more modular shaders, I'd like to just be _absolutely sure_ that there's no performance regressions here.

Are you able to please run the stress tests on a Quest 2 and just make sure we haven't lost any performance from this? If it's an improvement or no change - all good! But if there is a performance regression, then I'd much rather you just copy-pasta from Hotham's shaders 😄

File: hotham/src/systems/rendering.rs [hunk 1]
```diff
@@ -131,11 +129,11 @@ pub unsafe fn begin(
...
     for instanced_primitive in render_context.primitive_map.values() {
         let primitive = &instanced_primitive.primitive;
-        for (i, instance) in instanced_primitive.instances.iter().enumerate() {
+        for (instance, i) in instanced_primitive.instances.iter().zip(0u32..) {
             cull_data.push(&PrimitiveCullData {
-                index_instance: i.try_into().unwrap(),
-                index_offset: primitive.index_buffer_offset,
                 bounding_sphere: instance.bounding_sphere,
+                index_instance: i,
+                primitive_id: primitive.index_buffer_offset,
```
Comment (kanerogers) [inline hotham/src/systems/rendering.rs:136]: Good change, I like this 👍🏻

File: examples/custom-rendering/src/custom_render_context.rs [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+    vk,
+    vk_shader_macros::include_glsl,
+    Engine,
+};
+use std::{collections::HashMap, mem::size_of, slice};
+
+static QUADRIC_VERT: &[u32] = include_glsl!("src/shaders/quadric.vert", target: vulkan1_1);
+static QUADRIC_FRAG: &[u32] = include_glsl!("src/shaders/quadric.frag", target: vulkan1_1);
+
+// TODO: Ensure that this index doesn't collide with hotham engine internals.
```
[the remark this answers is not in the document]
Comment (rasmusgo) [reply examples/custom-rendering/src/custom_render_context.rs:16]: It does not! 😄 This used to be `6` and was inside the regular descriptor set but it has its own descriptor set now.

**Comment by rasmusgo** [comment]
There was a performance regression uncovered by the benchmark so I reverted the changes tp `pbr.frag` which resolved it. I will try to reduce the duplication later on but this works for now.

**Review by kanerogers — APPROVED** [review]
This is good stuff!

I am a little nervous about making more of the renderer public, but I suppose it's a reasonable stop gap before exposing a cleaner custom rendering API. 

Approved! 🎉

**Pull request merged**